### ISSUE TITLE
Difference between tflite int8 cpu inference and edgetpu inference

### ISSUE THREAD

**Issue by maxlacourchristensen** [google-coral/edgetpu#382]
I am having trouble matching tflite int8 cpu inference results with edgetpu results. To demonstrate this, I took your 
retrain_classification_ptq_tf2.ipynb example code and reduced the model to just the first Conv2D layer.

1) The first script creates the keras model, converts to tflite with quantization, runs inference on an image and saves the inferred upon image (example_image.npy) and the result (tflite_output.npy) to disk.

2) I then use the edgetpu_compiler version 15.0.340273435 to create the tflite for coral.

3) I upload that tflite to coral together with example_image.npy and run the second script.

4) The second script outputs the result: coral_output.npy which i transfer back to my desktop

5) The third script loads the tflite int8 result and the coral results and compares. Here i get a max absolute difference of 5.56.

What did I do wrong? 

Best regards,
Max

First script:

```python
import tensorflow as tf
assert float(tf.__version__[:3]) >= 2.3

import os
import numpy as np
import matplotlib.pyplot as plt
import cv2

_URL = "https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz"

zip_file = tf.keras.utils.get_file(origin=_URL,
                                   fname="flower_photos.tgz",
                                   extract=True)

flowers_dir = os.path.join(os.path.dirname(zip_file), 'flower_photos')

IMAGE_SIZE = 224
BATCH_SIZE = 64

datagen = tf.keras.preprocessing.image.ImageDataGenerator(
    rescale=1./255,
    validation_split=0.2)

train_generator = datagen.flow_from_directory(
    flowers_dir,
    target_size=(IMAGE_SIZE, IMAGE_SIZE),
    batch_size=BATCH_SIZE,
    seed=1337,
    subset='training')

val_generator = datagen.flow_from_directory(
    flowers_dir,
    target_size=(IMAGE_SIZE, IMAGE_SIZE),
    batch_size=BATCH_SIZE,
    seed=1337,
    subset='validation')

print (train_generator.class_indices)

labels = '\n'.join(sorted(train_generator.class_indices.keys()))

with open('flower_labels.txt', 'w') as f:
  f.write(labels)

IMG_SHAPE = (IMAGE_SIZE, IMAGE_SIZE, 3)


model = tf.keras.applications.MobileNetV2(input_shape=IMG_SHAPE,
                                              include_top=False,
                                              weights='imagenet')

model = tf.keras.Model(model.input, model.layers[2].output)
model.summary()

print('Number of trainable weights = {}'.format(len(model.trainable_weights)))

converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

with open('mobilenet_v2_1.0_224.tflite', 'wb') as f:
  f.write(tflite_model)

# A generator that provides a representative dataset
def representative_data_gen():
  dataset_list = tf.data.Dataset.list_files(flowers_dir + '/*/*', seed=1337)
  print(dataset_list)
  for i in range(100):
    image = next(iter(dataset_list))
    image = tf.io.read_file(image)
    image = tf.io.decode_jpeg(image, channels=3)
    image = tf.image.resize(image, [IMAGE_SIZE, IMAGE_SIZE])
    image = tf.cast(image / 255., tf.float32)
    image = tf.expand_dims(image, 0)
    yield [image]

converter = tf.lite.TFLiteConverter.from_keras_model(model)
# This enables quantization
converter.optimizations = [tf.lite.Optimize.DEFAULT]
# This sets the representative dataset for quantization
converter.representative_dataset = representative_data_gen
# This ensures that if any ops can't be quantized, the converter throws an error
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
# For full integer quantization, though supported types defaults to int8 only, we explicitly declare it for clarity.
converter.target_spec.supported_types = [tf.int8]
# These set the input and output tensors to uint8 (added in r2.3)
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8
tflite_model = converter.convert()

with open('mobilenet_v2_1.0_224_quant.tflite', 'wb') as f:
  f.write(tflite_model)

batch_images, batch_labels = next(val_generator)

def set_input_tensor(interpreter, input):
  input_details = interpreter.get_input_details()[0]
  tensor_index = input_details['index']
  input_tensor = interpreter.tensor(tensor_index)()[0]
  # Inputs for the TFLite model must be uint8, so we quantize our input data.
  # NOTE: This step is necessary only because we're receiving input data from
  # ImageDataGenerator, which rescaled all image data to float [0,1]. When using
  # bitmap inputs, they're already uint8 [0,255] so this can be replaced with:
  #   input_tensor[:, :] = input
  scale, zero_point = input_details['quantization']
  input_tensor[:, :] = np.uint8(input / scale + zero_point)

def classify_image(interpreter, input):
  set_input_tensor(interpreter, input)
  interpreter.invoke()
  output_details = interpreter.get_output_details()[0]
  output = interpreter.get_tensor(output_details['index'])
  # Outputs from the TFLite model are uint8, so we dequantize the results:
  scale, zero_point = output_details['quantization']
  output = scale * (output - zero_point)
  return output

interpreter = tf.lite.Interpreter('mobilenet_v2_1.0_224_quant.tflite')
interpreter.allocate_tensors()

cv2.imshow('image', batch_images[0])
cv2.waitKey(-1)

np.save('./example_image.npy', batch_images[0])

prediction = classify_image(interpreter, batch_images[0])
print(np.amin(prediction), np.amax(prediction))

np.save('./tflite_output.npy', prediction)
```

Second script:

```python
import numpy as np
import platform
import time
import glob
import cv2
import os

import tflite_runtime.interpreter as tflite

EDGETPU_SHARED_LIB = {
  'Linux': 'libedgetpu.so.1',
  'Darwin': 'libedgetpu.1.dylib',
  'Windows': 'edgetpu.dll'
}[platform.system()]

def make_interpreter(model_file):
    model_file, *device = model_file.split('@')
    return tflite.Interpreter(model_path=model_file,
                              experimental_delegates=[
                              tflite.load_delegate(EDGETPU_SHARED_LIB,
                              {'device': device[0]} if device else {})])

def set_input_tensor(interpreter, input):
  input_details = interpreter.get_input_details()[0]
  tensor_index = input_details['index']
  input_tensor = interpreter.tensor(tensor_index)()[0]
  # Inputs for the TFLite model must be uint8, so we quantize our input data.
  # NOTE: This step is necessary only because we're receiving input data from
  # ImageDataGenerator, which rescaled all image data to float [0,1]. When using
  # bitmap inputs, they're already uint8 [0,255] so this can be replaced with:
  #   input_tensor[:, :] = input
  scale, zero_point = input_details['quantization']
  #input_tensor[:, :] = np.uint8(input / scale + zero_point)
  input_tensor[:, :] = np.int8(input / scale + zero_point)

def classify_image(interpreter, input):
  set_input_tensor(interpreter, input)
  interpreter.invoke()
  output_details = interpreter.get_output_details()[0]
  output = interpreter.get_tensor(output_details['index'])
  # Outputs from the TFLite model are uint8, so we dequantize the results:
  scale, zero_point = output_details['quantization']
  output = scale * (output - zero_point)
  return output

interpreter = make_interpreter('./mobilenet_v2_1.0_224_quant_edgetpu.tflite')
interpreter.allocate_tensors()

inp = np.load('./example_image.npy')

prediction = classify_image(interpreter, inp)
print(np.amin(prediction), np.amax(prediction))

np.save('coral_output.npy', prediction)
```

Third script:

```python
import numpy as np

tflite = np.load('tflite_output.npy')
coral = np.load('coral_output.npy')

diff = tflite-coral
idx = np.abs(diff) > 1e-6
print(np.sort(np.abs(diff[idx])))

np.testing.assert_allclose(tflite, coral, rtol=1e-03, atol=1e-05)
```

**Comment by maxlacourchristensen** [comment]
Hi Naveen, any update on this? 
Best, Max

**Comment by Naveen-Dodda** [comment]
@maxlacourchristensen I see that you we comparing all the probable predictions from the model. Though there are few difference in values, do this effect the real semantic meaning of the output ?

I mean does these differences effect the real meaning or prediction of class.

**Comment by maxlacourchristensen** [comment]
Hi Naveen,

Notice that I reduced the model to just the first Conv2D layer to show the difference in the output and to reduce the number of places the difference can originate from. It is not comparing probable predictions. It is simply comparing the output from a single Conv2D layer. This error (max absolute difference of 5.56) is in my opinion too big for many applications. It will propagate through the network. Maybe for a classifier, such a big error can be tolerated but for our model of interest which is a human pose keypoint model the end result is almost a halving of the accuracy (compared to running tflite int8 cpu inference). Can you advise if we are doing something wrong?

In my opinion, semantic meaning of a neural network does not make sense. It boils down to mathematical operators (mult+add) which can be implemented to give bitwise the same output as a naive c++ implementation of the same layers (for example Conv2D). Can it really be true, that coral does not implement a Conv2D correctly or am I doing something wrong?  

Best,
Max

**Comment by Naveen-Dodda** [comment]
@maxlacourchristensen 

Thanks very much for debugging any pointing out with in Conv layer. We understand that this would influence position of the pose in your application. Can you identify if the (max absolute difference of 5.56) is consistent each time of inference at similar position in the matrix or with variable input can you see if the max difference occurs only at specific area.
